crates/stackable-operator/src/commons/networking.rs

```diff
@@ -1,44 +1,136 @@
-use std::{fmt::Display, ops::Deref};
+use std::{fmt::Display, net::IpAddr, ops::Deref, str::FromStr};
 
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
+use snafu::Snafu;
 
 use crate::validation;
 
-/// A validated hostname type, for use in CRDs.
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+/// A validated domain name type conforming to RFC 1123, so e.g. not an IP addresses
+#[derive(
+    Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd, JsonSchema,
+)]
 #[serde(try_from = "String", into = "String")]
-pub struct Hostname(#[validate(regex(path = "validation::RFC_1123_SUBDOMAIN_REGEX"))] String);
+pub struct DomainName(#[validate(regex(path = "validation::RFC_1123_SUBDOMAIN_REGEX"))] String);
+
+impl FromStr for DomainName {
+    type Err = validation::Errors;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        validation::is_rfc_1123_subdomain(value)?;
+        Ok(DomainName(value.to_owned()))
+    }
+}
 
-impl TryFrom<String> for Hostname {
+impl TryFrom<String> for DomainName {
     type Error = validation::Errors;
 
     fn try_from(value: String) -> Result<Self, Self::Error> {
-        validation::is_rfc_1123_subdomain(&value)?;
-        Ok(Hostname(value))
+        value.parse()
     }
 }
 
-impl From<Hostname> for String {
-    fn from(value: Hostname) -> Self {
+impl From<DomainName> for String {
+    fn from(value: DomainName) -> Self {
         value.0
     }
 }
 
-impl Display for Hostname {
+impl Display for DomainName {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.write_str(&self.0)
     }
 }
 
-impl Deref for Hostname {
+impl Deref for DomainName {
     type Target = str;
 
     fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
+#[derive(Debug, Snafu)]
+pub enum HostNameParseError {
+    #[snafu(display(
+        "the given hostname {hostname:?} is not a valid hostname, which needs to be either a domain name or IP address"
+    ))]
+    InvalidHostname { hostname: String },
+}
+
+/// A validated hostname, which is either a [`DomainName`] or [`IpAddr`].
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
+#[serde(try_from = "String", into = "String")]
+pub enum HostName {
+    IpAddress(IpAddr),
+    DomainName(DomainName),
+}
```

Comment:
Do we actually care about distinguishing these cases?

Reply:
I personally modeled it as String, but @Techassi suggested using an enum instead, which sounded fine to me. No preference from me.

Reply:
The RFC calls for these two distinct variants of a hostname.

Why should we remove all the value we get from using proper types by just using a plain string? We _could_ simplify the `DomainName` variant to use `String`, but I would definitely keep the enum.

Reply:
Ultimately the question here is the purpose of the type. Detailed types makes sense for inspecting and manipulating the hostname. The original intent for `Hostname` was the much smaller scope of "make sure this opaque looks close enough to a hostname that it won't fuck up the syntax of the config file it's string-templated into".

Reply:
> looks close enough

If we want to achieve this, sure. But then why reference RFC 1123 and only vaguely conform to it. If we want to conform to various RFCs, we should do so as closely as possible.

Another region where we _try_ to conform to an RFC (1035), we also completely miss the mark when validating domain labels. So I'm just trying to avoid the same thing happening here (even tho the domain name is not 100% valid).